pkg/config/store.go

```diff
@@ -0,0 +1,107 @@
+package config
+
+import (
+	"fmt"
+	"strings"
+)
+
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// ChangeType denotes the type of change
+type ChangeType int
+
+const (
+	// ChangeTypeUpdate - change was an update or a create to a key.
+	ChangeTypeUpdate ChangeType = iota
+	// ChangeTypeDelete - key was removed.
+	ChangeTypeDelete
+)
+
+// Change - A change record in the changelog.
+type Change struct {
+	// Key that was affected
+	Key string `json:"key"`
+	// Type how did the key change
+	Type ChangeType `json:"change_type"`
+	// change log index number of the change
+	Index int `json:"index"`
+}
+
+// KVStore defines the back end interface used by mixer
+// and the mixer config API server
+// It should support back ends like redis, etcd and NFS
+// All commands should return an index which can be used
+// to Read changes. If a KVStore does not support it,
+// it may return -1
+type KVStore interface {
+	// Get value at a key, false if not found.
+	Get(key string) (value string, index int, found bool)
+
+	// Set a value.
+	Set(key string, value string) (index int, err error)
+
+	// List keys with the prefix.
+	List(key string, recurse bool) (keys []string, index int, err error)
+
+	// Delete a key.
+	Delete(key string) error
+}
+
+// StoreListener listens for calls from the store that some keys have changed.
+type StoreListener interface {
+	// StoreChange notify listener that a new change is available
+	StoreChange(index int)
```

Comment:
`StoreChange` is called by `KVStore` when a new change is available in the underlying store. It will only be notified of committed changes. 
The KVStore client then goes back and issues `ReadChangeLog`, after it's last processed index number. 

If mixer crashes, It goes thru the cache initialization process. 
1. List Config    (this returns the `index` when the list operation was started)
2. Read all the listed keys and build cache
3. Register ChangeListener.
4. `ReadChangeLog` from `index` and update cache. Update `index`
5. When change listener is called essentially repeat step 4 with the currently processed index.

Change listener will be on a watchdog, so that we won't miss change, `ReadChangeLog` can be issues any time.